### DISCUSSION TITLE
How to add new packages to offline Verdaccio server

### DISCUSSION THREAD

**Discussion by michaelfelixmurphy** [verdaccio/verdaccio#2306] [Q&A]
Hello. First of all, thank you for this excellent tool. I've been impressed with the quality of Verdaccio and its docs. I have a question that despite about 10+ hours of research and trial and error, I have not been able to figure out.

# Background

I have a Verdaccio instance running on an entirely offline machine (an "air-gapped" network). I am using that Verdaccio instance as an offline npm mirror, so that I can build npm projects in the offline network. When I first set up this server, I cached all my projects' dependencies in a Verdaccio `storage` directory. I did this by running a Verdaccio instance on an online machine with registry.npmjs.org as an uplink. I ran all the `npm install`s I needed to, then copied the populated `storage` directory over to my offline Verdaccio instance.

# Problem

Now I need to add new packages into the offline Verdaccio instance. I need to somehow download packages on an online machine, and then copy them over to the offline Verdaccio instance. It seems I cannot do what I did initially. I can't get a new `storage` directory and copy it into the current offline Verdaccio instance. If I did that, I'd lose all the packages I'm currently serving there. Instead it seems I need to "merge" my existing `storage` directory with a new one from my online machine that just contains my new packages.

* **Main Question**: How do you recommend accomplishing this task?

I have tried a few things, which I'll lay out below, but I'm happy to abandon those ideas for whatever you recommend as the easiest/best approach.

The primary reason this is tricky is I might be adding new versions of a package where I already have some versions of that package cached in my offline Verdaccio instance. Let's take Bootstrap as an example. Say I have on my offline Verdaccio instance a `bootstrap/package.json` that lists Bootstrap versions up to 4.5.0 and a `bootstrap-4.4.0.tgz` tarball. What if I want to add Bootstrap 4.5.0? Presumably the `package.json` file needs to be updated. What if I want to add Bootstrap 5? Definitely the `package.json` file needs to be updated.

# Attempted Solution

I figured the best way to accomplish this would be to run another Verdaccio instance offline to use as an "importer" uplink. I'd add my `importer` Verdaccio instance as the only uplink for my primary offline Verdaccio instance. The `importer`'s `storage` would be temporary, so I could totally wipe it out and replace it with a new `storage` containing my new packages. Then I'd `npm install <new package names>` on one of my offline machines. That would request them from the `primary` Verdaccio instance, who would say "I don't have these" and then request them from the `importer` Verdaccio instance.

This works for package names that have never been cached on the `primary` Verdaccio instance. However for packages that already have *some* versions on the `primary` Verdaccio instance, it does not.

Let's take the Bootstrap example again. The version 4.4.0 tarball is cached on `primary` along with the `package.json` that I got from registry.npmjs.org through Verdaccio when I downloaded my initial `storage` cache. That `package.json` has an entry for Bootstrap 4.5.0. I've copied over a `storage` directory to `importer` that contains the 4.5.0 tarball and the `package.json` that I got just now when installed 4.5.0 on an online machine. (The `storage` directory also contains any dependencies - it was generated by doing a 4.5.0 install through an online Verdaccio instance).

### Problem

When I try to `env npm_config_repository=primary_verdaccio_instance npm install bootstrap` I get a 500 response. I have debugged this quite a bit. `primary` asks `importer` for `/bootstrap`. That request succeeds, though `primary` doesn't seem to do much with the information. Then my npm client asks `primary` for `bootstrap-4.5.0.tgz`. `primary` tries to find that file locally and fails. Then it looks in `bootstrap/package.json` and finds a registry.npmjs.org link. It tries to get the 4.5.0 tarball from that link and fails (because it can't reach the internet). It then returns 500 saying the uplink is offline. It never tries to reach `importer` who does in fact have the tarball.

* **Secondary Question**: If you think the above is the right approach, how do I make it work? How do I tell Verdaccio to only ever use the proxies/uplinks in my config file and *never* use the URLs in the `package.json` file?

  I saw [here](https://github.com/verdaccio/verdaccio/issues/1193#issuecomment-476074950) that the behavior where Verdaccio uses the `tarball` links in `package.json` is for backward compatibility so there may not be a way to disable it. If so, that's fine. Then I go back to my main question: what do you recommend instead?

Please let me know what information I can provide to help you answer this question. I'm happy to clarify what I'm asking or to provide any files/logs/etc that would help you debug my attempted solution.

Thanks!

**Comment by juanpicado** [comment]
hi @michaelfelixmurphy , I'm trying to understand the examples, maybe you can rework them which are more easy to see the differences.  The approach of having 2 it's fine, actually the first instance should take whatever is on the second one and the second one whatever  has cached from internet. Anyway, I'll try to read this again and maybe you can elaborate it a bit more so I'm sure I understood correctly, maybe I have some ideas in mind.

**Reply by michaelfelixmurphy** [reply]
Hi @juanpicado, thanks for getting back to me. I really appreciate any help I can get on this. And thanks for reading that wall of text. I know my first post was a bit long :)

I'll clarify my example about the issue I'm currently seeing, but as I said above, if you think there's a better approach to use, I'm happy to abandon my current plan and do what you recommend. My goal is simply: **how do I copy the files for new packages into an existing, completely offline Verdaccio server?**

## Example

I have two Verdaccio servers containing the following. Both are disconnected from the internet.

### primary

- `storage`
  - *other dependencies*
  - `bootstrap`
    - `bootstrap-4.4.0.tgz`
    - `package.json` (cached from registry.npmjs.org when 4.5.0 was the latest release, but my project needed 4.4.0)
    - ```json
      {
        ...
        "versions": {
          ...
          "4.4.0": ...,
          "4.5.0": {
            "dist": {
              ...
              "tarball": "https://registry.npmjs.org/bootstrap/-/bootstrap-4.5.0.tgz",
              ...
            },
            ...
          },
          ...
        },
        "distfiles": {
          ...
          "bootstrap-4.4.0.tgz": ...,
          "bootstrap-4.5.0.tgz": {
            "url": "https://registry.npmjs.org/bootstrap/-/bootstrap-4.5.0.tgz",
          },
          ...
        },
        ...
      }
      ```

### importer

- `storage`
  - *other dependencies*
  - `bootstrap`
    - `bootstrap-4.5.0.tgz`
    - `package.json` (cached from registry.npmjs.org just now when I requested 4.5.0; contents not relevant, but similar to the above)

I run `env npm_config_repository=primary npm install bootstrap`. I should be able to get bootstrap 4.5.0 from `importer`. And 4.5.0 should be cached into `primary` so that I no longer need `importer` to install 4.5.0 in the future.

What actually happens: `primary` asks `importer` for `/bootstrap`. That request succeeds, though `primary` doesn't seem to do much with the information. Then my npm client asks `primary` for `bootstrap-4.5.0.tgz`. `primary` tries to find that file locally and fails. Then it looks in `primary/storage/bootstrap/package.json`. It finds one of the above `4.5.0` entries and finds the `https://registry.npmjs.org/bootstrap/-/bootstrap-4.5.0.tgz` link. I'm not sure if it's finding that link from `versions/dist` or `distfiles`.

`primary` tries to download `https://registry.npmjs.org/bootstrap/-/bootstrap-4.5.0.tgz`. It fails because it's intentionally, permanently offline. `primary` never asks `importer` for `bootstrap-4.5.0.tgz`. If it did, it would succeed and all my worries would go away.

Neither `primary` nor `importer` have `registry.npmjs.org` as an uplink or a proxy. In fact the only uplink/proxy configured is that `importer` is set as an uplink and proxy for all packages in `primary`'s config.

If the above is the right approach, then my question is: **How do I tell Verdaccio to only ever use the proxies/uplinks in my config file and never use the URLs in the package.json file?**

Please let me know if I can further clarify anything.

**Reply by michaelfelixmurphy** [reply]
Hi @juanpicado, do you have any suggestions on how to get Verdaccio to contact another Verdaccio instance for specific package versions instead of going to the registry.npmjs.org link in packages' package.json file?

Any information you can provide would be greatly appreciated. My clarified example is above. And please do let me know if there's anything more I can do to help you understand my question.

Thanks!

**Reply by juanpicado** [reply]
I see, I think I got it, btw, verdaccio uses `__distfiles` as reference. Anyway I kinda replicated the test you describe and works for me. This is the schema I've used.

![issue_2306](https://user-images.githubusercontent.com/558752/126878822-04d6819c-aaee-4ebe-8104-5a0031a9a574.png)

Anyway, I still see lack of information. I'd recommend you post here the following things.

- Project package.json example
- Verdaccio configuration of primary
- Verdacccio configuration of importer
- Steps, 1,2,3,4,5 with npm commands.
- Share what you see on the server log
- Share what you have on the each storage
- **ENSURE** both verdaccio uses different `./storage` folder, if you use both on the same config, the server will crash (due one of them lock the storage and other one is not able to read.
